internal/datacoord/services.go

```diff
@@ -1070,3 +1071,62 @@ func getDiff(base, remove []int64) []int64 {
 	}
 	return diff
 }
+
+// AcquireSegmentLock acquire the reference lock of the segments.
+func (s *Server) AcquireSegmentLock(ctx context.Context, req *datapb.AcquireSegmentLockRequest) (*commonpb.Status, error) {
+	resp := &commonpb.Status{
+		ErrorCode: commonpb.ErrorCode_UnexpectedError,
+	}
+
+	if s.isClosed() {
+		log.Warn("failed to acquire segments reference lock for closed server")
+		resp.Reason = msgDataCoordIsUnhealthy(Params.DataCoordCfg.GetNodeID())
+		return resp, nil
+	}
+
+	hasSegments := s.meta.HasSegments(req.SegmentIDs)
+	if !hasSegments {
+		resp.Reason = "segment does not exist"
+		return resp, nil
+	}
+
+	err := s.segRefer.AddSegmentsLock(req.SegmentIDs, req.NodeID)
+	if err != nil {
+		log.Warn("Add reference lock on segments failed", zap.Int64s("segIDs", req.SegmentIDs), zap.Error(err))
+		resp.Reason = err.Error()
+		return resp, nil
+	}
+	hasSegments = s.meta.HasSegments(req.SegmentIDs)
+	if !hasSegments {
+		err = retry.Do(ctx, func() error {
+			return s.segRefer.ReleaseSegmentsLock(req.SegmentIDs, req.NodeID)
+		}, retry.Attempts(100))
+		resp.Reason = "segment does not exist"
```

Comment:
if all failed, should directly panic here. actually I don't recommend to add retry here, all retry and panic should happened etcd kv, but so far it's ok.